java-checks/src/main/java/org/sonar/java/checks/CollectorsToListCheck.java

```diff
@@ -91,13 +92,30 @@ protected void onMethodInvocationFound(MethodInvocationTree mit) {
     } else {
       return;
     }
-    Symbol assignedVariable = findAssignedVariable(mit.parent());
-    if (!mutable || assignedVariable == null
-      || assignedVariable.usages().stream().noneMatch(CollectorsToListCheck::isListBeingModified)) {
+    if (isInvariantTypeArgument(mit) && (!mutable || isReturnedListUnmodified(mit))) {
       reportIssue(collector, mutable);
     }
   }
 
+  private static boolean isReturnedListUnmodified(MethodInvocationTree mit) {
+    Symbol assignedVariable = findAssignedVariable(mit.parent());
+    return assignedVariable == null || assignedVariable.usages().stream().noneMatch(CollectorsToListCheck::isListBeingModified);
+  }
+
+  private static boolean isInvariantTypeArgument(MethodInvocationTree mit) {
+    Type ownerType = mit.methodSymbol().owner().type();
+    if (ownerType.isRawType()) {
+      return true;
+    }
+    Type toListType = mit.symbolType();
+    if (toListType.isRawType()) {
+      return false;
+    }
+    Type toListArgType = toListType.typeArguments().get(0);
+    Type ownerArgType = ownerType.typeArguments().get(0);
+    return toListArgType.is(ownerArgType.fullyQualifiedName());
+  }
```

Comment:
Looking again at it, and thanks to your comment, I feel we can simplify it even more and make it more explicit!
1. If the `Stream` is NOT a raw type, it's semantically impossible to have the `collect` method returning a raw type. So we can skip the second test, we will simply never match it.
2. I changed the names of variables because I made a mistake in my previous suggestion (we are reacting on `collect`, not on `toList`) and we can make it explicit. The owner is a `Stream` also, so let's try to be as close as possible to the reality of the code.

```suggestion
  private static boolean isInvariantTypeArgument(MethodInvocationTree collectMethodInvocation) {
    Type streamType = collectMethodInvocation.methodSymbol().owner().type();
    if (streamType.isRawType()) {
      return true;
    }
    Type collectArgType = collectMethodInvocation.symbolType().typeArguments().get(0);
    Type streamArgType = streamType.typeArguments().get(0);
    return collectArgType.is(streamArgType.fullyQualifiedName());
  }
```